python/graphstorm/dataloading/sampler.py

```diff
@@ -243,3 +246,105 @@ def _generate(self, g, eids, canonical_etype):
         dst = self._local_neg_nids[vtype][dst]
         dst = np.tile(dst, self.k)
         return th.as_tensor(src), th.as_tensor(dst)
+
+class FastMultiLayerNeighborSampler(NeighborSampler):
+    """ Fast MultiLayerNeighborSampler
+
+        If mask is None, it acts the same as dgl.dataloading.MultiLayerNeighborSampler
+
+    Parameters
+    ----------
+    reverse_edge_types_map: dict
+        A map for reverse edge type
+    """
+    def __init__(
+        self,
+        fanouts,
+        edge_dir="in",
+        prob=None,
+        mask=None,
+        replace=False,
+        prefetch_node_feats=None,
+        prefetch_labels=None,
+        prefetch_edge_feats=None,
+        output_device=None,
+        reverse_edge_types_map=None,
+    ):
+        self.mask = mask
+        self.reverse_edge_types_map = reverse_edge_types_map
+        self.original_edge_types_map = {
+            val: key for key, val in reverse_edge_types_map.items()
+        } if reverse_edge_types_map is not None else None
+
+        super().__init__(
+            fanouts=fanouts,
+            edge_dir=edge_dir,
+            prob=prob,
+            mask=None, # Do neighbor sampling with out edge mask
+            replace=replace,
+            prefetch_node_feats=prefetch_node_feats,
+            prefetch_labels=prefetch_labels,
+            prefetch_edge_feats=prefetch_edge_feats,
+            output_device=output_device
+        )
+
+    def sample_blocks(self, g, seed_nodes, exclude_eids=None):
+        output_nodes = seed_nodes
+        blocks = []
+        for fanout in reversed(self.fanouts):
+            frontier = g.sample_neighbors(
+                seed_nodes,
+                fanout,
+                edge_dir=self.edge_dir,
+                prob=self.prob,
+                replace=self.replace,
+                output_device=self.output_device,
+                exclude_edges=exclude_eids,
+            )
+            eid = frontier.edata[EID]
+            new_eid = dict(eid)
+            if self.mask is not None:
+                new_edges = {}
+                for etype in frontier.canonical_etypes:
+                    if self.mask in g.edges[etype].data:
+                        # train mask in data
+                        if etype in eid:
+                            mask = g.edges[etype].data[self.mask][eid[etype]].bool()
+                            new_edges[etype] = mask
+                            new_eid[etype] = eid[etype][mask]
```

Comment:
is it possible that there are no edges left after the filtering?

Reply:
In theory yes. But with large enough fanout, e.g., 5 or 10 and small val and test sets e.g.., < 10% 5%, the change is very low.